ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala

```diff
@@ -115,9 +115,9 @@ private[platform] class MutableCacheBackedContractStore(
   )(implicit
       loggingContext: LoggingContext
   ): Either[Set[ContractId], (Set[Timestamp], Set[ContractId])] = {
-    val cacheQueried = ids.map(id => id -> contractsCache.get(id))
+    val cacheQueried = ids.view.map(id => id -> contractsCache.get(id)).toVector
```

Comment:
this works if we want to prevent repeated hashCode/equals calculations for the set, but the original problem might be that the hashCode/equals calculation is expensive and not memoized.
Can it be that the memoization is just a missing piece? For example AFAICT the toString method is memoized, and also could be the basis for hashCode / equals.
@remyhaemmerle-da WDYT?

Reply:
Moreover: the StateCache for contract keys is also using the same hashCode/euqlas calculation

Reply:
Will will add a cache for hashCode in `com.daml.lf.value.Value.ContractId`

Reply:
@remyhaemmerle-da Could you say what your current thinking about this issue is? 

I guess the removal of the initial comment is related to the fact that the the problem is actually about the hashCode of the ContractStateValue rather than the ContractId, isn't it.

Reply:
(Sorry I have been interrupted when I was writing my comment)

After looking at the code, am I am a not the memoization of `ContractId`'s  `hashCode` plays a role here.
I mean, `com.daml.lf.Value.ContractId.V1` is a case class whose two fields have a cached `hashCode`, so the cost of the current `hashCode` should be reasonable.

I run some micro benchmarks and i measure a 10% speedup by caching the `hashCode` when converting an `Array[ContractId]` of  to  a `Set[ContractId]`. If you think it is really worthwhile I am fine to add it.

Reply:
If this above is true, why would we have hash calculation on heatmap, if the Set of ContractId-s are already constructed on the input? Then already all the ContractId-s should have their hashes memoized, no? Or the equals is the problem?  :thinking:

Reply:
@remyhaemmerle-da Indeed, for `ContractId` there is probably no cost. A potential issue, as @nmarton-da pointed out is with the `GlobalKey` which is used for lookups in the mutable contract state cache (which has a `Value` that can have prohibitively expensive `hashCode` in extreme cases) - see https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala#L64

Reply:
> If this above is true, why would we have hash calculation on heatmap, if the Set of ContractId-s are already constructed on the input? Then already all the ContractId-s should have their hashes memoized, no? Or the equals is the problem?

The problem is not with the `ContractId` but with the `ContractStateValue.Active.contract.arg` that is also a part of the tuple created by that map: 
```
    val cacheQueried = ids.map(id => id -> contractsCache.get(id))
```